daemon/containerd/image.go

```diff
@@ -341,3 +348,26 @@ func (i *ImageService) getAllImagesWithRepository(ctx context.Context, ref refer
 	nameFilter := "^" + regexp.QuoteMeta(ref.Name()) + ":" + reference.TagRegexp.String() + "$"
 	return i.client.ImageService().List(ctx, "name~="+strconv.Quote(nameFilter))
 }
+
+func (i *ImageService) getImageLabelByDigest(ctx context.Context, target digest.Digest, labelKey string) (string, error) {
+	imgs, err := i.client.ImageService().List(ctx, "target.digest=="+target.String())
+	if err != nil {
+		return "", err
+	}
+
+	var value string
+	for _, img := range imgs {
+		if v, ok := img.Labels[labelKey]; ok {
+			if value != "" && value != v {
+				log.G(ctx).WithFields(log.Fields{
+					"firstValue": value,
+					"nextValue":  v,
+				}).Warnf("conflicting value of label %s", labelKey)
```

Comment:
This block could use some explanation (as a comment) do outline the logic here.

- We list all images with a specific digest (IIUC, that's because compared to graph-driver, storage is "pivoted"; digests can occur multiple times (if an image is tagged with multiple names)
- Those images _can_ have a `imageLabelClassicBuilderParent` label associated
- If _multiple_ have that label, but they don't match, we print a warning, but we pick the first one

I'm curious here;

- first of all, would `i.client.ImageService().List()` allow filtering on the presence of a `org.mobyproject.image.parent` label? (if so, should we only return images that do?)
- second; would it be possible for different images to be used as parent for different builds? (in that case, is it actually an _error_ / _warning_ ?)
- if we _do_ have multiple results.. wondering if we should return a parent at all (because it's ambiguous, so we may be either returning the correct parent, or the wrong one?)

Reply:
> first of all, would i.client.ImageService().List() allow filtering on the presence of a org.mobyproject.image.parent label? (if so, should we only return images that do?)

It would be nice, but unfortunately all filters passed to `List` are `OR`'d, so we can't list images that BOTH match the digest AND have a label.

> second; would it be possible for different images to be used as parent for different builds? (in that case, is it actually an error / warning ?)

I don't think it should be possible, unless you really try hard and find a hash collision.

> if we do have multiple results.. wondering if we should return a parent at all (because it's ambiguous, so we may be either returning the correct parent, or the wrong one?)

In this case any of these would be a valid parent.

Reply:
> It would be nice, but unfortunately all filters passed to List are OR'd, so we can't list images that BOTH match the digest AND have a label.

I recalled that I looked at containerd filters at some point, and that there was a way to specify _AND_ conditions. The `List` function uses `ParseAll()`;  https://github.com/moby/moby/blob/ec259ddfeeb7d0e6b4a97ed6774a704f733cdb8f/vendor/github.com/containerd/containerd/filters/parser.go#L59-L70


And indeed, each _filter_ is combined with `OR`. HOWEVER (but please test if I'm correct); each of those filter statements can _combine_ multiple conditions by having a comma-separated list of conditions; see the `Parse()` function; https://github.com/moby/moby/blob/ec259ddfeeb7d0e6b4a97ed6774a704f733cdb8f/vendor/github.com/containerd/containerd/filters/parser.go#L27-L49


So in this case, something like this would (hopefully!) work to return images which digest matches, and that have a `org.mobyproject.image.parent` label (with any value);

```
target.digest=="+target.String()+",labels."+labelKey
```

Reply:
Good find, thanks!